### PR TITLE
Support serial data types.

### LINKED ISSUE #96: Support serial data types
Support `smallserial` `serial` and `bigserial` sql types see https://www.postgresql.org/docs/12/datatype-numeric.html

It is possible to do by updating `SqlType::SmallInt`, `SqlType::Integer` and `SqlType::BigInt` respectively by extending enum variants with `min` possible value field

### PR DESCRIPTION
**Pull request by suhaibaffan**
Closes #96
Please give your input @alex-dukhno!

### CODE REVIEW

**Comment by suhaibaffan** [comment]
Got it, thanks @alex-dukhno.

**Comment by suhaibaffan** [comment]
@alex-dukhno let me know if I am on the right direction or not?

File: src/sql_types/src/lib.rs [hunk 1]
```diff
@@ -96,14 +96,20 @@ pub trait Serializer {
...
+    min: u16,
+}
 
 impl Constraint for SmallIntTypeConstraint {
     fn validate(&self, in_value: &str) -> Result<(), ConstraintError> {
-        match lexical::parse::<i16, _>(in_value) {
-            Ok(_) => Ok(()),
-            Err(e) if e.code == lexical::ErrorCode::InvalidDigit => Err(ConstraintError::NotAnInt),
-            Err(_) => Err(ConstraintError::OutOfRange),
+        if self.min < lexical::parse::<u16, _>(in_value).unwrap() {
```
Comment (alex-dukhno) [inline src/sql_types/src/lib.rs:105]: Could you please move `if` check inside `Ok(parsed)` branch, `unwrap`ing will panic and `in_value` will be parsed only once

File: src/sql_types/src/lib.rs [hunk 2]
```diff
@@ -124,14 +130,20 @@ impl Serializer for SmallIntTypeSerializer {
...
+    min: u32,
+}
 
 impl Constraint for IntegerSqlTypeConstraint {
     fn validate(&self, in_value: &str) -> Result<(), ConstraintError> {
-        match lexical::parse::<i32, _>(in_value) {
-            Ok(_) => Ok(()),
-            Err(e) if e.code == lexical::ErrorCode::InvalidDigit => Err(ConstraintError::NotAnInt),
-            Err(_) => Err(ConstraintError::OutOfRange),
+        if self.min < lexical::parse::<u32, _>(in_value).unwrap() {
```
Comment (alex-dukhno) [inline src/sql_types/src/lib.rs:139]: same as above

File: src/sql_types/src/lib.rs [hunk 3]
```diff
@@ -152,14 +164,20 @@ impl Serializer for IntegerSqlTypeSerializer {
...
+}
 
 impl Constraint for BigIntTypeConstraint {
     fn validate(&self, in_value: &str) -> Result<(), ConstraintError> {
-        match lexical::parse::<i64, _>(in_value) {
-            Ok(_) => Ok(()),
-            Err(e) if e.code == lexical::ErrorCode::InvalidDigit => Err(ConstraintError::NotAnInt),
-            Err(_) => Err(ConstraintError::OutOfRange),
+        if self.min < lexical::parse::<u64, _>(in_value).unwrap() {
+            match lexical::parse::<u64, _>(in_value) {
```
Comment (alex-dukhno) [inline src/sql_types/src/lib.rs:174]: same as above

File: src/sql_types/src/lib.rs (hunk 1 above)
Comment (suhaibaffan) [reply src/sql_types/src/lib.rs:105]: Sure, thanks.

**Comment by alex-dukhno** [comment]
You can check `clippy` errors with `./ci/clippy-local.sh`
`cargo test --lib --all` - unit tests and
`pytest -v tests/functional/*` - functional end-to-end tests

**Comment by suhaibaffan** [comment]
@alex-dukhno In `sql_engine/src/ddl/create_table`,
`sqlparser::ast::DataType::Regclass => SqlType::Integer(1),` I can use [Regclass](https://docs.rs/sqlparser/0.5.1/sqlparser/ast/enum.DataType.html#variant.Regclass) for just Serial, what do you suggest for `smallSerial` and `bigSerial`?

**Comment by alex-dukhno** [comment]
I am not sure about `Regclass` ... I bet that `Custom(ObjectName)` is what you are looking for https://github.com/ballista-compute/sqlparser-rs/blob/main/src/ast/data_type.rs#L67
But I can be wrong 😄

File: src/sql_engine/src/ddl/create_table.rs [hunk 4]
```diff
@@ -46,12 +46,15 @@ impl<P: BackendStorage> CreateTableCommand<P> {
...
-                        sqlparser::ast::DataType::SmallInt => SqlType::SmallInt,
-                        sqlparser::ast::DataType::Int => SqlType::Integer,
-                        sqlparser::ast::DataType::BigInt => SqlType::BigInt,
+                        sqlparser::ast::DataType::SmallInt => SqlType::SmallInt(i16::min_value()),
+                        sqlparser::ast::DataType::Int => SqlType::Integer(i32::min_value()),
+                        sqlparser::ast::DataType::BigInt => SqlType::BigInt(i64::min_value()),
                         sqlparser::ast::DataType::Char(len) => SqlType::Char(len.unwrap_or(255)),
                         sqlparser::ast::DataType::Varchar(len) => SqlType::VarChar(len.unwrap_or(255)),
                         sqlparser::ast::DataType::Boolean => SqlType::Bool,
+                        sqlparser::ast::DataType::Custom(ObjectName(_serial)) => SqlType::Integer(1),
```
Comment (alex-dukhno) [inline src/sql_engine/src/ddl/create_table.rs:55]:
```suggestion
                        sqlparser::ast::DataType::Custom(name) => {
                            let name = name.to_string();
                            if name == "smallserial" {
                                SqlType::SmallInt(1);
                            } else if name == "serial" {
                                SqlType::Integer(1);
                            } else if name == "bigserial" {
                                SqlType::BigInt(1);
                            } else {
                                unimplemented!("{:?} type is not supported", name)
                            }
                        },
```

Comment (alex-dukhno) [reply src/sql_engine/src/ddl/create_table.rs:55]: this should help

Comment (suhaibaffan) [reply src/sql_engine/src/ddl/create_table.rs]: Thanks  @alex-dukhno fixed, please review it.

**Comment by alex-dukhno** [comment]
I'll merge it as I am currently working on type improvements

**Comment by alex-dukhno** [comment]
Thanks for the PR

**Pull request merged**